### PR TITLE
Kill plugin script

### PR DESCRIPTION
**Pull request by danwinship**
The CNI port already got us most of the way there. This moves the remaining bits of pod setup/teardown into go rather than shell script.

(I started this a long time ago, and finished it now because of #11990.)

@openshift/networking PTAL

### CODE REVIEW

File: pkg/sdn/plugin/pod_linux.go [hunk 1]
```diff
@@ -254,20 +190,97 @@ func (m *podManager) ipamDel(id string) error {
...
+		otx.AddFlow("table=7, priority=100, ip, nw_dst=%s, actions=output:%d", podIP, ofport)
+	} else {
+		otx.AddFlow("table=7, priority=100, reg0=0, ip, nw_dst=%s, actions=output:%d", podIP, ofport)
+		otx.AddFlow("table=7, priority=100, reg0=%d, ip, nw_dst=%s, actions=output:%d", vnid, podIP, ofport)
+	}
+
+	return otx.EndTransaction()
+}
+
+func (m *podManager) setupPodBandwidth(pod *kapi.Pod, hostVeth string) error {
```
Comment (smarterclayton) [inline pkg/sdn/plugin/pod_linux.go]: Are these methods actually related?  If all you're accessing is `m.ovs` taking ovs as an argument is just as good.  We generally lean from a style perspective on not creating grouping types unless there is an actual "group".

**Comment by danwinship** [comment]
@smarterclayton like that?

File: pkg/sdn/plugin/pod_linux.go [hunk 2]
```diff
@@ -254,20 +191,88 @@ func (m *podManager) ipamDel(id string) error {
...
+		otx.AddFlow("table=7, priority=100, ip, nw_dst=%s, actions=output:%d", podIP, ofport)
+	} else {
+		otx.AddFlow("table=7, priority=100, reg0=0, ip, nw_dst=%s, actions=output:%d", podIP, ofport)
+		otx.AddFlow("table=7, priority=100, reg0=%d, ip, nw_dst=%s, actions=output:%d", vnid, podIP, ofport)
+	}
+
+	return otx.EndTransaction()
+}
+
+func setupPodBandwidth(ovsif *ovs.Interface, pod *kapi.Pod, hostVeth string) error {
```
Comment (smarterclayton) [inline pkg/sdn/plugin/pod_linux.go]: Yeah, looks good.  I find a lot of the time it's easy to "over OO" stuff, and I think for many functions you end up with more testable and less coupled code.

**Comment by dcbw** [comment]
@danwinship So generally LGTM, though we'll have the same upgrade problem when a master+node host is updated, since the RPM changes remove the script.  Then pods on that node will fail because the script doesn't exist.  Maybe we just keep the script around in the RPM (I'm pretty sure we don't  need it in the images, just in the specfile) for one more release?

**Comment by danwinship** [comment]
> Maybe we just keep the script around in the RPM (I'm pretty sure we don't need it in the images, just in the specfile) for one more release?

openshift doesn't care where the script is as long as it's in $PATH, so we could have the update process copy it from /usr/bin to /usr/local/bin before upgrading the RPM, and then delete it after restarting the node service

**Comment by danwinship** [comment]
(re-push is just to squash the fixup)

**Comment by smarterclayton** [comment]
Generally if it's more complex to do it in ansible, doing it in your code
(preserving the script for at least one version) reduces the impact of this
change to other teams (keeps your dirty laundry secret)

**Comment by knobunc** [comment]
Do we only support single version upgrades?  I.e. to get from 3.2 to 3.4
you need to go through 3.3 (and start it up)?

**Comment by danwinship** [comment]
> (keeps your dirty laundry secret)

It's not clear whose dirty laundry it is; leaving openshift running while changing the rpm out from under it is kind of dirty on their part. But sure, we can leave the script around.

> Do we only support single version upgrades?

yes

**Comment by smarterclayton** [comment]
Not quite true.  We support single version no downtime upgrades.  We do not
support multi version no downtime (of the control plane) upgrades, although
anything beyond the control plane needs an exception.

On Dec 7, 2016, at 12:22 PM, Dan Winship <notifications@github.com> wrote:

(keeps your dirty laundry secret)

It's not clear whose dirty laundry it is; leaving openshift running while
changing the rpm out from under it is kind of dirty on their part. But
sure, we can leave the script around.

Do we only support single version upgrades?

yes

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/openshift/origin/pull/12145#issuecomment-265511764>,
or mute
the thread

**Comment by danwinship** [comment]
[test][testextended][extended:networking]

**Comment by danwinship** [comment]
Are we going to try to get this in for 3.5? (It's not a feature so I guess it's not subject to the "feature freeze" deadline anyway, but...)

**Comment by danwinship** [comment]
Rebased and repushed now that the tree is open for 3.6. Also changed it to make it continue installing the now-unused script, to avoid problems during upgrade. We can remove it completely for 3.7.

**Comment by danwinship** [comment]
flake #12855, [testextended]

**Comment by danwinship** [comment]
@openshift/networking the test failures will be fixed by #13061 when it lands. This is ready for review.

File: pkg/sdn/plugin/pod_linux.go [hunk 3]
```diff
@@ -252,20 +191,83 @@ func (m *podManager) ipamDel(id string) error {
...
+	// note pod ingress == OVS egress and vice versa, and OVS ingress is in Kbps
+	if podIngress != nil {
+		ovsEgress = podIngress.Value()
+	}
+	if podEgress != nil {
+		ovsIngress = podEgress.Value() / 1024
+	}
+
+	if ovsEgress > 0 {
+		qos, err := ovsif.Create("qos", "type=linux-htb", fmt.Sprintf("other-config:max-rate=%d", ovsEgress))
```
Comment (pravisankar) [inline pkg/sdn/plugin/pod_linux.go]: Where is the code equivalent to "ip link set dev ${veth_host} qlen 1000" in openshift-sdn-ovs?

File: pkg/sdn/plugin/pod_linux.go [hunk 4]
```diff
@@ -252,20 +191,83 @@ func (m *podManager) ipamDel(id string) error {
...
+		return nil
+	}
+
+	var ovsIngress, ovsEgress int64
+	// note pod ingress == OVS egress and vice versa, and OVS ingress is in Kbps
+	if podIngress != nil {
+		ovsEgress = podIngress.Value()
+	}
+	if podEgress != nil {
+		ovsIngress = podEgress.Value() / 1024
```
Comment (pravisankar) [inline pkg/sdn/plugin/pod_linux.go]: is it /1024 or /1000? 1024 might be right (we are using 1000 in openshift-sdn-ovs)

**Comment by pravisankar** [comment]
Added couple of comments, rest of the code LGTM.
Need to resolve some merge conflicts

File: pkg/sdn/plugin/pod_linux.go (hunk 3 above)
Comment (danwinship) [inline pkg/sdn/plugin/pod_linux.go]: ah, i think this branch originally predates that addition and I must have missed it when rebasing

File: pkg/sdn/plugin/pod_linux.go (hunk 4 above)
Comment (danwinship) [inline pkg/sdn/plugin/pod_linux.go]: Oh, yeah, I saw 1000 in openshift-sdn-ovs but assumed it was probably wrong and should be 1024. @dcbw?

**Comment by danwinship** [comment]
Rebased and added in the qlen thing (and did another check to see if I'd missed anything when rebasing, and didn't see anything). It seems that the netlink library used for all the other /sbin/ip stuff in pod_linux.go doesn't let you do "set qlen" (even if we updated to the latest git) so I had to use ipcmd instead.

File: pkg/sdn/plugin/pod_linux.go (hunk 4 above)
Comment (danwinship) [inline pkg/sdn/plugin/pod_linux.go]: ovs docs say "ingress_policing_rate: ... Maximum rate for data received on this interface, in kbps." so 1024 is correct

**Comment by danwinship** [comment]
[merge]
